File: chap10_cellregulation/motif_FF.m
%  -------------------------------------------------------------------
%
%   The feedforward motif.
%
%   For Chapter 10, Section 10.1.3 of
%   Keener and Sneyd, Mathematical Physiology, 3rd Edition, Springer.
%
%   Written by James Keener and James Sneyd.
%
%  -------------------------------------------------------------------

function motif_FF

clear all
close all
clc
set(0,                           ...
   'defaultaxesfontsize', 20,   ...
   'defaultaxeslinewidth', 2.0, ...
   'defaultlinelinewidth', 2.0);
init=[0 0]
tspan=linspace(0,50,5000);

[T,Y] = ode15s(@rhs,tspan,init);
R = Y(:,1);
X = Y(:,2);
S = (T>10) + (T>20) +  (T>30);

plot(T,R)
ylabel('R')
yyaxis('right')
plot(T,S)
ylabel('S')
xlabel('t')
legend('boxoff')
legend('R','S','location','southeast')
set(gca,'linewidth',1.5)
box off
out = [T R S];
%save('test.dat','out','-ascii')

end % of main


%%
function out = rhs(t,Y)
R = Y(1);
X = Y(2);

n=2;
K=2;
S =  (t>10) +  (t>20) + (t>30);
phi = K*X^n/(K+X^n);
%phi = X; % for exercise 1.9
out(1) = S - phi*R;
out(2) = S - X;
out = out';

end

File: chap10_cellregulation/motif_FF.py

#   -------------------------------------------------------------------
# 
#    The feedforward motif.
# 
#    For Chapter 10, Section 10.1.3 of
#    Keener and Sneyd, Mathematical Physiology, 3rd Edition, Springer.
#  
#    Written by James Keener and James Sneyd.
#  
#   ------------------------------------------------------------------- 

import numpy as np
import matplotlib.pyplot as plt
from scipy.integrate import solve_ivp

# Parameters
n = 2
K = 2
init = [0, 0]

# Define the ODE system
def rhs(t, Y):
    R, X = Y
    S = (t > 10) + (t > 20) + (t > 30)
    phi = K * X**n / (K + X**n)
    dRdt = S - phi * R
    dXdt = S - X
    return [dRdt, dXdt]

# Time points
tspan = np.linspace(0, 50, 5000)

# Integrate the ODEs
sol = solve_ivp(rhs, [tspan[0], tspan[-1]], init, t_eval=tspan)

# Extract results
T = sol.t
R = sol.y[0]
X = sol.y[1]
S = (T > 10).astype(int) + (T > 20).astype(int) + (T > 30).astype(int)

# Plot the results
plt.plot(T, R, label='R')
plt.plot(T, S, label='S')
plt.xlabel('Time')
plt.ylabel('Concentration')
plt.legend()
plt.grid(True)
plt.show()
